### PR TITLE
Add ONLY_SERIALIZE for NLOHMANN_DEFINE_DERIVED_TYPE_* macros

### PR DESCRIPTION
**Pull request by EinarsNG**
Added additional macros to expand on the #4033 pull request.
2 new macros:
- `NLOHMANN_DEFINE_DERIVED_TYPE_INTRUSIVE_ONLY_SERIALIZE`
- `NLOHMANN_DEFINE_DERIVED_TYPE_NON_INTRUSIVE_ONLY_SERIALIZE`

Works identically to the ones added in the pull request mentioned above except only allows serialization.

* * *

## Pull request checklist

Read the [Contribution Guidelines](https://github.com/nlohmann/json/blob/develop/.github/CONTRIBUTING.md) for detailed information.

- [x]  Changes are described in the pull request, or an [existing issue is referenced](https://github.com/nlohmann/json/issues).
- [x]  The test suite [compiles and runs](https://github.com/nlohmann/json/blob/develop/README.md#execute-unit-tests) without error.
- [x]  [Code coverage](https://coveralls.io/github/nlohmann/json) is 100%. Test cases can be added by editing the [test suite](https://github.com/nlohmann/json/tree/develop/test/src).
- [x]  The source code is amalgamated; that is, after making changes to the sources in the `include/nlohmann` directory, run `make amalgamate` to create the single-header files `single_include/nlohmann/json.hpp` and `single_include/nlohmann/json_fwd.hpp`. The whole process is described [here](https://github.com/nlohmann/json/blob/develop/.github/CONTRIBUTING.md#files-to-change).

### CODE REVIEW

File: docs/mkdocs/docs/api/macros/nlohmann_define_derived_type.md [hunk 1]
```diff
@@ -1,20 +1,23 @@
...
 ```
 
 These macros can be used to simplify the serialization/deserialization of derived types if you want to use a JSON
 object as serialization and want to use the member variable names as object keys in that object.
 
-- Macros 1 and 2 are to be defined **inside** the class/struct to create code for.
+- Macros 1, 2 and 3 are to be defined **inside** the class/struct to create code for.
 Like [`NLOHMANN_DEFINE_TYPE_INTRUSIVE`](nlohmann_define_type_intrusive.md), they can access private members.
-- Macros 3 and 4 are to be defined **outside** the class/struct to create code for, but **inside** its namespace.
+- Macros 3, 4 and 5 are to be defined **outside** the class/struct to create code for, but **inside** its namespace.
```
Comment (nlohmann) [inline docs/mkdocs/docs/api/macros/nlohmann_define_derived_type.md]: I think you mean 4, 5, and 6.

File: docs/mkdocs/docs/api/macros/nlohmann_define_derived_type.md [hunk 2]
```diff
@@ -48,14 +51,20 @@ friend void to_json(nlohmann::json&, const type&);
 friend void from_json(const nlohmann::json&, type&);
 ```
 
-Macros 3 and 4 add two functions to the namespace which take care of the serialization and deserialization:
+Macros 4 and 5 add two functions to the namespace which take care of the serialization and deserialization:
```
Comment (nlohmann) [inline docs/mkdocs/docs/api/macros/nlohmann_define_derived_type.md:54]: Maybe also 6?

**Review by nlohmann — CHANGES_REQUESTED** [review]

File: docs/mkdocs/docs/api/macros/nlohmann_define_derived_type.md (hunk 2 above)
Comment (EinarsNG) [reply docs/mkdocs/docs/api/macros/nlohmann_define_derived_type.md:54]: `add two functions`
3 and 6 only add one.
I have mentioned that one a little bit below this line:
`Macros 3 and 6 add one function to the namespace which take care of the serialization only:`

**Review by nlohmann — APPROVED** [review]
Looks good to me.

**Comment by nlohmann** [comment]
Thanks a lot!

**Pull request merged**